pkg/cluster/platformworkloadidentities.go

```
@@ -42,7 +44,11 @@ func (m *manager) platformWorkloadIdentityIDs(ctx context.Context) error {
 
 		identityDetails, err := m.userAssignedIdentities.Get(ctx, resourceId.ResourceGroupName, resourceId.Name, &armmsi.UserAssignedIdentitiesClientGetOptions{})
 		if err != nil {
-			return fmt.Errorf("error occured when retrieving platform workload identity '%s' details: %w", operatorName, err)
+			if azureerrors.IsInvalidIdentityUserError(err) {
+				return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidPlatformWorkloadIdentity, operatorName, fmt.Sprintf("platform workload identity '%s' is invalid", resourceId.Name))
```

The one thing we're losing here is visibility into the actual error message that's here.  

Should we minimize the errors we're checking for, or log out the error message internally?  If we limit it to 404s on the identity, we know we're hitting it, but if all of a sudden we start getting 418s we're going to lose it in the response.

---

I know we should / could assume any 400 is going to be a user error, but we could hit an issue internally where we're forming a bad GET request to the user identity, which is an RP error rather than a customer error.  Hence me raising it.

---

> we could hit an issue internally where we're forming a bad GET request to the user identity

this is why I'm not fully comfortable with just saying "any 4xx error here is a UserError". I'd rather handle the specific, obvious cases we know about explicitly as UserErrors here, and log the rest and return an internal server error, and we can then add future enhancements for any new surprises we find out in the field later. 

Perhaps we should just always log the error here before determining how to handle it?

---

@bennerv 

>but we could hit an issue internally where we're forming a bad GET request to the user identity, which is an RP error rather than a customer error. Hence me raising it.

After doing some CIF works, my perspective on this became a little bit more pragmatic. Firstly, I agree that we shouldn't lose the specific error message. In reality, however, I've seen more UserErrors (4**) caused by users, anecdotally 99% of the time, the rest are mostly MSFT service issues, and very seldom due to an RP bug. Because of this observation. Let's say it's 80/20, I believe it's more practical to return a UserError by default because there is relatively little chance that they're actually ServerErrors such as caused by RP or MSFT service, these servers are programmed to not fail although they're not perfect. On the other hand if ServerErrors are returned by default, there is a high chance that there are actually UserErrors, users are prone to make mistakes anyways and this was very evident in our CIF dashboards in the past. One of the stories I did was properly converting the ServerErrors that was skewing our reporting to UserErrors, there were a lot of them.

In the end sending ServerError when it's not and UserError when it's not both upsets a customer, I just believe that the former is very less likely. Hence I recommended just returning a generic UserError plus the original error message. WDYT?

---

> @pepedocs I think what you've suggested makes sense, but it is a deviation from the current strategy we've generally been using to handle errors in the RP.
> 
> For the sake of not blocking this individual PR on such a shift, I'd suggest we let this PR merge as-is, since this is a fairly high-priority change, and tackle our error handling more generally in a separate, focused effort, in which we get more eyes and hands on the desired outcome and corresponding changes.

Agree. @slawande2 Could you please create a spike jira to investigate/continue this discussion?

---

Can we at least log out the error message internally so if we see something that's not a 404 or common error (it's actually us causing the issue) we know what is going on?

---

Yeah there is no justification for hiding the error by not logging at the very least. Your on-call and SRE will thank you.